### PR TITLE
[StateService] get historical state

### PR DESCRIPTION
**Pull request by ZhangTao1596**
Changes:
    * Add `GetState` `FindStates` API to get historical storage or list storages by prefix
    * Add `README` 
@devhawk @shargon

### CODE REVIEW

File: src/StateService/Storage/StateStore.cs
Comment (shargon) [inline src/StateService/Storage/StateStore.cs:58]: ?

File: src/StateService/StatePlugin.cs
Comment (devhawk) [inline src/StateService/StatePlugin.cs:260]: It seems like this should be two separate Rpc Methods - one for get and one for find - instead of using a parameter to change behavior

File: src/StateService/StatePlugin.cs
Comment (devhawk) [inline src/StateService/StatePlugin.cs:299]: There needs to be a mechanism to retrieve records that were truncated in the first request. possibly an optional `page number` parameter which will skip over `max` * `page number` records

**Review by devhawk — CHANGES_REQUESTED** [review]
comments left in GetState

File: src/StateService/StatePlugin.cs
Comment (ZhangTao1596) [reply src/StateService/StatePlugin.cs]: Done.

File: src/StateService/StatePlugin.cs
Comment (ZhangTao1596) [reply src/StateService/StatePlugin.cs]: Done.

File: src/StateService/Storage/StateStore.cs
Comment (ZhangTao1596) [reply src/StateService/Storage/StateStore.cs]: Removed.

**Comment by roman-khimov** [comment]
I think returning state without proofs is not a very good thing. And to get a single KV pair we can already use `getstateroot` with `getproof`, so only problem seems to be a set of KV pairs that of course can be served by `getstate` but
 * accepting stateroot hash instead of block index/hash (it's trivial to get via `getstateroot` and it's needed for subsequent checks anyway)
 * without `IsFind` parameter (assuming it to always be true)
 * always returning a set of KV pairs along with boundary proofs (for the first and the last pairs)
This way we could always not just get some data from RPC node, but actually verify it and trust this data correctness.

File: src/StateService/README.md
Comment (devhawk) [inline src/StateService/README.md:45]: docs need to be updated to match implementation

File: src/StateService/StatePlugin.cs
Comment (devhawk) [inline src/StateService/StatePlugin.cs:288]: What is the rationale for specifying a max page number?

File: src/StateService/StateService/config.json
Comment (devhawk) [inline src/StateService/StateService/config.json:8]: What happens if there is a find query with over 100x100 results?

**Comment by devhawk** [comment]
> I think returning state without proofs is not a very good thing.

Given that we're not removing the ability to get proofs, what's the concern @roman-khimov? Are you worried about incenting developers to trust remote notes w/o proofs? Instead of a separate method for providing state w/o proof, could we have a method that returns the proof and the state, so that it can be retrieved in one round trip rather than requiring every key to require two round trips (getproof + verifyproof) 

> always returning a set of KV pairs along with boundary proofs (for the first and the last pairs) This way we could always not just get some data from RPC node, but actually verify it and trust this data correctness.

I love this idea. Since the data may span multiple pages, is it sufficient to return the boundary proofs for the first and last pairs *on the page being returned*?

**Comment by roman-khimov** [comment]
> Are you worried about incenting developers to trust remote notes w/o proofs?

Yeah, it's not the end of the world thing, but still we better return things that can be checked when we can.

>  is it sufficient to return the boundary proofs for the first and last pairs on the page being returned?

Yep, that should work, one can reconstruct a part of the MPT from a set of KV pairs and boundary proofs, per-page proofs work just fine for this too.

**Comment by devhawk** [comment]
> Yeah, it's not the end of the world thing, but still we better return things that can be checked when we can.

OK, so let's change the new getstate method to return *both* the proof and the value. I'd rather add a new method and not break the existing getproof method (though I suspect the number of users of that API at this moment is very low)

File: src/StateService/StateService/config.json
Comment (ZhangTao1596) [reply src/StateService/StateService/config.json:8]: This is the default settings, if you want more, you can change it.

**Comment by ZhangTao1596** [comment]
> * always returning a set of KV pairs along with boundary proofs (for the first and the last pairs)
>   This way we could always not just get some data from RPC node, but actually verify it and trust this data correctness.

If you are suspicious of `getstate` results,  you can use `getproof` to verify. `getstate` is a extension of `getstorage`.

File: src/StateService/README.md
Comment (ZhangTao1596) [reply src/StateService/README.md:45]: Will do.

File: src/StateService/StatePlugin.cs
Comment (ZhangTao1596) [reply src/StateService/StatePlugin.cs:288]: User can use Empty as prefix and scan all storages, Considering there is no memory protection mechanism in MPT impl, it's dangerous to resolve all branches.

File: src/StateService/README.md
Comment (ZhangTao1596) [reply src/StateService/README.md:45]: Updated.

**Comment by roman-khimov** [comment]
> I don't think it's necessary to return both KV and proof, this makes result very large.

You don't need to return proof for every KV, only the first and the last element in a set (page) that you return, it's not that hard and StateService has all the data anyway.

File: src/StateService/StatePlugin.cs
Comment (roman-khimov) [reply src/StateService/StatePlugin.cs:272]: IIUC it's contract-bound anyway, although NEO or GAS contracts can of course have a lot of data. At the same time, walking through whole NEO or GAS storage doesn't mean storing whole MPT for them in memory, there is no need to (reminds me of nspcc-dev/neo-go#2019 and the way it throws away from memory things it doesn't need).

But it can be improved even further if we're to add `from` path into request, this way you can start resolving `from` (not iterating whole `prefix` until you end up on a requested page) and then walk the number of elements needed and return them. The next request will have new `from`.

**Comment by ZhangTao1596** [comment]
> only the first and the last element in a set (page) that you return,

Why proofs of the first and last can prove all results reliable?

**Comment by roman-khimov** [comment]
> Why proofs of the first and last can prove all results reliable?

Consider this simplified trie and a request for prefix 2:
s

You trust root hash and you get yellow proofs in return along with all green KV pairs. If you have all green KV pairs (that's important, they need to be sequential and there can't be any omissions in a sequence, but for prefix-based search that's exactly what we have) you can easily reconstruct this part of MPT and compare yellow paths, that will prove all of this subtrie (because you have a hash of 212 in 21 and you get 212 hash by adding 2121, etc).

**Comment by ZhangTao1596** [comment]
> > Why proofs of the first and last can prove all results reliable?
> 
> Consider this simplified trie and a request for prefix 2:
> s
> 
> You trust root hash and you get yellow proofs in return along with all green KV pairs. If you have all green KV pairs (that's important, they need to be sequential and there can't be any omissions in a sequence, but for prefix-based search that's exactly what we have) you can easily reconstruct this part of MPT and compare yellow paths, that will prove all of this subtrie (because you have a hash of 212 in 21 and you get 212 hash by adding 2121, etc).

So user can verify results only using the very first proof and the very last proof cross pages.  We don't need to add proof in every page. We have `getproof`.

File: src/StateService/StatePlugin.cs
Comment (ZhangTao1596) [reply src/StateService/StatePlugin.cs:278]: Use `from` is better.

Comment (roman-khimov) [reply src/StateService/StatePlugin.cs:278]: This technically also allows for user-defined paging (specifying number of results to return).

**Comment by roman-khimov** [comment]
Then consumer of this API will have to retrieve and store all pages before he can verify the result. It's not very productive IMO and some users of this API might as well stop iterating at some point. I'd prefer proving data as it comes (in pages).

File: src/StateService/StatePlugin.cs
Comment (ZhangTao1596) [reply src/StateService/StatePlugin.cs]: Updated.

**Comment by ZhangTao1596** [comment]
I still don't like both proofs and KV in one request. We have `getproof`. If so, the method should be percisely `FindStatesWithProof`. Should we add proof in `getstate`?

**Comment by devhawk** [comment]
> I still don't like both proofs and KV in one request. We have `getproof`. If so, the method should be percisely `FindStatesWithProof`. Should we add proof in `getstate`?

I'm *personally* fine with having `getstate` not include the proof since we also have `getproof`/`verifyproof`. If we were designing in a greenfield, I'd argue for a single RPC method that returned storage item + proof, but since we have shipped the current version I think adding a `getstate` without proof makes the most sense.

> So user can verify results only using the very first proof and the very last proof cross pages. We don't need to add proof in every page. We have `getproof`

Part of the point is to retrieve information in as few round trips as possible. For `findstate` I'd rather get the proof for the first and last element of each page than require `findstate`, `getproof` and `verifyproof`

**Review by devhawk — APPROVED** [review]

File: src/StateService/StatePlugin.cs
Comment (devhawk) [reply src/StateService/StatePlugin.cs]: This is a nice improvement. Great suggestion @roman-khimov

**Comment by ZhangTao1596** [comment]
@superboyiii Hi bro. Can you test this pr?

**Comment by devhawk** [comment]
@shargon can you give this PR a look?

File: src/StateService/MPT/MPTTrie.Find.cs
Comment (shargon) [inline src/StateService/MPT/MPTTrie.Find.cs:87]: Linq again

File: src/StateService/MPT/MPTTrie.Find.cs
Comment (shargon) [inline src/StateService/MPT/MPTTrie.Find.cs:106]: Avoid Linq

File: src/StateService/MPT/MPTTrie.Find.cs
Comment (shargon) [inline src/StateService/MPT/MPTTrie.Find.cs:122]: Avoid Linq

File: src/StateService/StatePlugin.cs
Comment (shargon) [inline src/StateService/StatePlugin.cs:253]: I prefer to avoid JArray, and receive the arguments. It will be easy to know what should I send to `FindState`

File: src/StateService/StatePlugin.cs
Comment (shargon) [inline src/StateService/StatePlugin.cs:291]: Does you tested with a big tree if we could have a DoS here?

**Comment by shargon** [comment]
could you add also https://github.com/neo-project/neo-modules/issues/635 ?

**Comment by ZhangTao1596** [comment]
> could you add also #635 ?

Will do.

File: src/StateService/StateService/config.json
Comment (ZhangTao1596) [reply src/StateService/StateService/config.json:7]: We don't use pages anymore. We use `from`.

File: tests/Neo.Plugins.StateService.Tests/MPT/UT_MPTTrie.cs
Comment (ZhangTao1596) [reply tests/Neo.Plugins.StateService.Tests/MPT/UT_MPTTrie.cs:588]: Mistake.

File: src/StateService/StatePlugin.cs
Comment (ZhangTao1596) [reply src/StateService/StatePlugin.cs:291]: I will. But the values are all contract bound. Besides with `from` we only walk through part of tree.

File: tests/Neo.Plugins.StateService.Tests/MPT/UT_MPTTrie.cs
Comment (ZhangTao1596) [reply tests/Neo.Plugins.StateService.Tests/MPT/UT_MPTTrie.cs]: Updated.

File: src/StateService/MPT/MPTTrie.Find.cs
Comment (ZhangTao1596) [reply src/StateService/MPT/MPTTrie.Find.cs]: Done.

File: src/StateService/MPT/MPTTrie.Find.cs
Comment (ZhangTao1596) [reply src/StateService/MPT/MPTTrie.Find.cs]: Updated.

File: src/StateService/MPT/MPTTrie.Find.cs
Comment (ZhangTao1596) [reply src/StateService/MPT/MPTTrie.Find.cs]: Updated.

File: src/StateService/StatePlugin.cs
Comment (ZhangTao1596) [reply src/StateService/StatePlugin.cs]: Updated.

File: src/StateService/MPT/MPTTrie.Find.cs
Comment (ZhangTao1596) [reply src/StateService/MPT/MPTTrie.Find.cs]: Updated.

File: src/StateService/MPT/MPTTrie.Find.cs
Comment (ZhangTao1596) [reply src/StateService/MPT/MPTTrie.Find.cs]: Updated.

File: src/StateService/StateService/config.json
Comment (shargon) [reply src/StateService/StateService/config.json:7]: Then we don't need https://github.com/neo-project/neo-modules/pull/638/files#diff-a2e7e45020d834da7043d6854062d505dbac22640129ce2e82803df4cdf59719R12

Comment (ZhangTao1596) [reply src/StateService/StateService/config.json:7]: Done.

Comment (ZhangTao1596) [reply src/StateService/StateService/config.json:7]: Done.

File: src/StateService/MPT/MPTTrie.Find.cs
Comment (shargon) [inline src/StateService/MPT/MPTTrie.Find.cs:106]: The condition `0 < from.Length` could be outside the `for`

Comment (ZhangTao1596) [reply src/StateService/MPT/MPTTrie.Find.cs]: Done. Please have a second review.

Comment (shargon) [reply src/StateService/MPT/MPTTrie.Find.cs]: @superboyiii please test it

File: src/StateService/StatePlugin.cs
Comment (shargon) [inline src/StateService/StatePlugin.cs:268]: Move this condition to line 257?

File: src/StateService/StatePlugin.cs
Comment (shargon) [inline src/StateService/StatePlugin.cs:308]: Move after parse `root_hash`?

File: src/StateService/MPT/MPTTrie.Find.cs
Comment (shargon) [inline src/StateService/MPT/MPTTrie.Find.cs:80]: Maybe we can receive a `CancellationToken` and break the iteration if the connection died?

File: src/StateService/StatePlugin.cs
Comment (devhawk) [inline src/StateService/StatePlugin.cs:298]: In addition to the data, you need to return some indicator if there are more items to be found

**Review by devhawk — CHANGES_REQUESTED** [review]
FindState RPC method needs a way of indicating if we are at the end of the array of keys or if there are more to retrieve

File: src/StateService/MPT/MPTTrie.Find.cs
Comment (devhawk) [inline src/StateService/MPT/MPTTrie.Find.cs:65]: If from key is specified, it should not be included in the results returned by Find.

**Comment by devhawk** [comment]
FYI, I set up a local neo-cli node for testnet running StateService compiled out of this branch. Worked like a charm the first time out!

File: src/StateService/StatePlugin.cs
Comment (ZhangTao1596) [reply src/StateService/StatePlugin.cs]: Done.

File: src/StateService/StatePlugin.cs
Comment (ZhangTao1596) [reply src/StateService/StatePlugin.cs:308]: Done.

File: src/StateService/StatePlugin.cs
Comment (ZhangTao1596) [reply src/StateService/StatePlugin.cs]: Done.

File: src/StateService/MPT/MPTTrie.Find.cs
Comment (ZhangTao1596) [reply src/StateService/MPT/MPTTrie.Find.cs:80]: Currently rpc method can't get `HttpContext`. I'm afraid this change will affect all rpc method.
Maybe I can create another pr.
Do you have any suggestion about get http context here without influence on other rpc methods?

File: src/StateService/MPT/MPTTrie.Find.cs
Comment (superboyiii) [reply src/StateService/MPT/MPTTrie.Find.cs]: > @superboyiii please test it

Works well.

File: src/StateService/StatePlugin.cs
Comment (superboyiii) [reply src/StateService/StatePlugin.cs]: > Did you tested with a big tree and result if we could have a DoS here?

100 items costs 5ms, 1000 items costs 25ms from my side. I think it's OK, not very cost time, much better than getNep17transfers.

**Comment by superboyiii** [comment]
Tested. Pass
Works well from my side.

File: src/StateService/MPT/MPTTrie.Find.cs
Comment (devhawk) [reply src/StateService/MPT/MPTTrie.Find.cs:65]: Note, this is still an issue as of latest commit in this PR (661284c08907608a93c7331fd90084de89af27a1)

File: src/StateService/MPT/MPTTrie.Find.cs
Comment (devhawk) [reply src/StateService/MPT/MPTTrie.Find.cs:80]: We should do this in a separate PR to address #629 generally, not just for state service

**Comment by devhawk** [comment]
Did @roman-khimov sign off on *not* including boundary proofs in FindStates?

File: src/StateService/MPT/MPTTrie.Find.cs
Comment (ZhangTao1596) [reply src/StateService/MPT/MPTTrie.Find.cs:80]: > We should do this in a separate PR to address #629 generally, not just for state service

Agree.

File: src/StateService/MPT/MPTTrie.Find.cs
Comment (ZhangTao1596) [reply src/StateService/MPT/MPTTrie.Find.cs:65]: I have not finished this fix. Will inform you then.

File: src/StateService/MPT/MPTTrie.Find.cs
Comment (ZhangTao1596) [reply src/StateService/MPT/MPTTrie.Find.cs]: Done.

File: src/StateService/MPT/MPTTrie.Find.cs
Comment (ZhangTao1596) [reply src/StateService/MPT/MPTTrie.Find.cs:65]: Done.

**Comment by ZhangTao1596** [comment]
> Did @roman-khimov sign off on _not_ including boundary proofs in FindStates?

Proof added.

**Comment by ZhangTao1596** [comment]
Merge?

**Review by devhawk — DISMISSED** [review]

**Review by shargon — APPROVED** [review]

**Comment by devhawk** [comment]
> Merge?

Yes please

**Comment by roman-khimov** [comment]
We can return an empty array of KV pairs and proof of nonexistence (trie path up to the point where it can't go further because of missing element).

**Comment by ZhangTao1596** [comment]
> Actually, one thing. If you call `FindStates` with an unused contract hash (like `UInt160.Zero`) you get an error `The given key was not present in the dictionary`. I was expecting an empty array of states. Throwing exceptions when there is no data doesn't seem like the right choice for state service RPC endpoints to me, but I'm interested in what @shargon and @roman-khimov think

It should be `Unkown contract`.

**Comment by ZhangTao1596** [comment]
> We can return an empty array of KV pairs and proof of nonexistence (trie path up to the point where it can't go further because of missing element).

We can't. If the contract is not deployed, There is no contract id. If we want to prove, we can prove the contract inexistent in contract management contract.

File: src/StateService/StatePlugin.cs
Comment (vncoelho) [inline src/StateService/StatePlugin.cs:267]: Maybe it should be returned as an exception because it could mislead the interpretation.

**Review by devhawk — CHANGES_REQUESTED** [review]
~How come ledger contract data is not included in state service? ~

Never mind

**Comment by roman-khimov** [comment]
Primary reasons for that are described in #520. But we have `getblock` and `getrawtransaction` for ledger things.

**Review by devhawk — APPROVED** [review]

File: src/StateService/StatePlugin.cs
Comment (ZhangTao1596) [reply src/StateService/StatePlugin.cs:267]: How? Can you describe this in detail? We have `truncated` in response and the results are ordered by key.

File: src/StateService/StatePlugin.cs
Comment (shargon) [inline src/StateService/StatePlugin.cs:260]: We should check the max length of this argument, ensure that is less than `MaxKeyLength`

Comment (ZhangTao1596) [reply src/StateService/StatePlugin.cs:260]: Done.

**Review by shargon — APPROVED** [review]

**Comment by superboyiii** [comment]
Tested Pass.

**Review by superboyiii — APPROVED** [review]

**Comment by superboyiii** [comment]
@shargon Could you review again?

**Comment by superboyiii** [comment]
@shargon Merge?

**Review by shargon — APPROVED** [review]

**Pull request merged**
